File: TobiiExampleScript/Tobii_helpers/t2t/examples/python/followEyes.py
#!/usr/bin/env python

"""
# Luca Filippin - April 2010 - luca.filippin@gmail.com                                                
# Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste. All rights reserved.
# 
# PLEASE HAVE A LOOK TO THE SWIG DOCUMENTATION FOR A DEEPER UNDERSTANDING OF THIS CODE
# NOTE: Swig wrapper to t2t makes use of shadow classes and swig helpers (see t2tsw.i)
"""

from argparse import ArgumentParser
import sys
import turtle
from t2tsw import *
from t2tHelpers import *


def banner():
    print ""
    print "+-------------------------------------------------------------------------------------+"
    print "|  This is a sample application which shows eyes movements as tracked by the Tobii ET |"
    print "|-------------------------------------------------------------------------------------|"
    print "|  Luca Filippin - April 2010 - luca.filippin@gmail.com                               |"                                                
    print "|  Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste     |"
    print "+-------------------------------------------------------------------------------------+"
    print ""

def parseOptions():
    parser = ArgumentParser(description='Display eyes movements by traking them through a Tobii Eyes Tracker')
    parser.add_argument('tobii_ip', help='tobii ip address')
    parser.add_argument('--version', action = 'version', version='%(prog)s v1.0')
    parser.add_argument("-l", "--log-file", dest="tobii_log", default="", help="log file name")
    parser.add_argument("-c", "--calibration-file", dest="calibration_file", default="", help="Use a stored calibration file")
    parser.add_argument("-p", "--port", dest="tobii_port", type=int, default=4455, help="TET server listening port")
    parser.add_argument("-d", "--debug-mode", dest="debug_mode", action="store_true", help="Print gaze samples")
    parser.add_argument("-m", "--camera", dest="camera", action="store_true", help="Use camera coordinates")
    parser.add_argument("-q", "--quick-start", dest="quick_start", action="store_true", help="Start soon after the command has been launched")
    parser.add_argument("-w", "--window-width", dest="window_width", type=float, default=0.5, help="Window width: range (0,1]")
    parser.add_argument("-e", "--window-height", dest="window_height", type=float, default=0.5, help="Window heigth: range (0,1]")
    parser.add_argument("-s", "--shape-size", dest="shape_size", type=float, default=0.4, help="size of the eyes: float > 0")
    parser.add_argument("-t", "--tracking-on", dest="tracking_on", action="store_true", help="Show the path made by the turtle")
    parser.add_argument("-v", "--validity", dest="validity", type=int, default=2, choices=xrange(0, 5), help='validity level: 0 = certainly, 1 = probably, 2 = 50%%, 3 = likely not 4 = surely not')
   
    try:
        o = parser.parse_args()
        if o.shape_size <= 0:
            raise Exception("Bad window shape size value '%(v)f': should be in (0, 1]" %{ 'v' : o.shape_size })
        if o.window_width <= 0 or o.window_width > 1:
            raise Exception("Bad window width value '%(v)f': should be in (0, 1]" %{ 'v' : o.window_width })
        if o.window_height <= 0 or o.window_height > 1:
            raise Exception("Bad window height value '%(v)f': should be in (0, 1]" %{ 'v' : o.window_height })
        if o.tobii_port < 0:
            raise Exception("Bad port value '%(v)s': should be > 0" %{ 'v' : o.tobii_port })
    except Exception, e:
        raise Exception("++ Argument error\n" + e.__str__())
     
    return o

def followEyes(eye_l, eye_r, line, validity, tracking, debug, camera, w, h, root):
    colors = ("red", "orange", "yellow", "blue", "white");
    
    c = t2tCmd()
    c.cmd = "GET_SAMPLE_EXT"    
    smp = cmdEx(c).sample_ext
    
    left = True
    right = True
	
    if camera:
	    lx, ly = 1 - smp.lxcam, smp.lycam
	    rx, ry = 1 - smp.rxcam, smp.rycam
    else:
	    lx, ly = smp.lx, smp.ly
	    rx, ry = smp.rx, smp.ry
	    
    if lx >= 0 and ly >= 0 and rx >= 0 and ry >= 0 and smp.lval <= validity and smp.rval <= validity:
        m = "Sample OO: " + smp.__str__()
        x, y = (lx + rx)*w/2, h-(ly + ry)*h/2
    elif rx >= 0 and  ry >= 0 and smp.rval <= validity:
        m = "Sample OX: " + smp.__str__()
        x, y = rx*w, h-ry*h
        left = False
    elif lx >= 0 and  ly >= 0 and smp.lval <= validity:
        m = "Sample XO: " + smp.__str__()
        x, y = lx*w, h-ly*h
        right = False
    else:
        m = "Sample XX: " + smp.__str__()
        right = False
        left = False
	
    if left:
        eye_l.showturtle()
        eye_l.color(colors[int(smp.lval)])
        eye_l.goto(lx*w, (1-ly)*h)
    else:
        eye_l.hideturtle()
            
    if right:
        eye_r.showturtle()
        eye_r.color(colors[int(smp.rval)])
        eye_r.goto(rx*w, (1-ry)*h)
    else:
        eye_r.hideturtle()
    
    if tracking:
        if left:
            eye_l.stamp()
        if right: 
            eye_r.stamp() 
        if left or right:
            line.goto(x, y)
                
        m += "\nMove to (x, y) = (%.0f, %.0f)" %(x, y)		
    else:
        m += "\nPrevious sample SKIPPED"
    
    if debug:
        print m;
    
    if not root.stop_follow_eyes:
        root.after(10, followEyes, eye_l, eye_r, line, validity, tracking, debug, camera, w, h, root)
    else:
        root.destroy()

def _get_out():
    root.stop_follow_eyes = True
    
try:

    banner()
    tobii_connected = False
    o = parseOptions() 
    if o.tobii_log <> "": 
        t2tOutputFileName(o.tobii_log, "w")
    
    print "Connecting to the TET server @" + o.tobii_ip + ":" + '%(p)d' %{ 'p':o.tobii_port } + "..."
    
    c = t2tCmd()
    c.cmd = "CONNECT"
    c.prm.connect.ip_address = o.tobii_ip
    c.prm.connect.port = o.tobii_port
    cmdEx(c)
    checkStatus("connect", 1, timeout = 5)
    
    if not checkStatus("connected", 1, timeout = 5): 
        bailOut(msg = "Connection failed\nExiting", disconnect = False)
    tobii_connected = True
    
    if o.calibration_file <> "":
        print "Loading calibration data " + o.calibration_file + "..."
        c = t2tCmd()
        c.cmd = "START_CALIBRATION"
        c.prm.start_calibration.load_from_file = 1
        c.prm.start_calibration.cmatrix.cols = 0
        c.prm.start_calibration.fname = o.calibration_file
        c.prm.start_calibration.cmatrix.rows = 0
        c.prm.start_calibration.cmatrix.vals = None
        cmdEx(c)
        
        checkStatus("calibrating", 1, timeout = 5)
        if not checkStatus("calibstarted", 1, timeout = 5): 
            bailOut(msg = "Start calibration failed\nExiting", disconnect = True)
        
        # there's no calibend flag set for loading
        if not checkStatus("calibrating", 0, timeout = 20): 
            bailOut(msg = "Start calibration failed\nExiting", disconnect = True)
    
    if o.quick_start:
        print "\n***** QUIT THE TRACKING WINDOW TO STOP *****\n"
    else:
        raw_input("\n***** PRESS <enter> TO START TRACKING, THEN QUIT THE TRACKING WINDOW TO STOP *****\n")
    
    c = t2tCmd()
    c.cmd = "START_TRACKING"
    cmdEx(c)
    checkStatus("running", 1, timeout = 5)
    if not checkStatus("runstarted", 1, timeout = 5):    
        bailOut(msg = "Start tracking failed\nExiting", disconnect = True)
    

    root = turtle.TK.Tk()
    w, h = root.winfo_screenwidth()*o.window_width, root.winfo_screenheight()*o.window_height
    root.title("Follow the eyes movements through a Tobii eyetracker")
    root.resizable(False, False)
    
    canvas = turtle.TK.Canvas(root, width=w, height=h, bg="#000000")
    canvas.pack()
    
    screen = turtle.TurtleScreen(canvas)
    screen.setworldcoordinates(0, 0, w, h)
    screen.bgcolor(0, 0, 0)
    
    eye_l = turtle.RawTurtle(screen)
    eye_r = turtle.RawTurtle(screen)
    
    for e in eye_l,eye_r:   
        e.shape("circle")
        e.color("red")
        e.shapesize(o.shape_size, o.shape_size, o.shape_size)
        e.speed(0)
        e.penup()
        e.hideturtle()
    
    line = turtle.RawTurtle(screen)
    line.color("green")
    line.speed(0)
    line.penup()
    line.hideturtle()
    
    if o.tracking_on:
        line.pendown();
 
    root.protocol( "WM_DELETE_WINDOW", _get_out)
    root.follow_eyes_id = []
    root.stop_follow_eyes = False
    followEyes(eye_l, eye_r, line, o.validity, o.tracking_on, o.debug_mode, o.camera, w, h, root)
    root.mainloop()
    
    c = t2tCmd()
    c.cmd = "STOP_TRACKING"
    cmdEx(c)
    checkStatus("stop", 1, timeout = 1)
    if not checkStatus("running", 0, timeout = 5): 
        bailOut(msg = "Stop tracking failed\nExiting", disconnect = True)
    
    tobii_connected = False
    c = t2tCmd()
    c.cmd = "DISCONNECT"
    cmdEx(c, disconnect = False)
    checkStatus("disconnect", 1, timeout = 5)
    if not checkStatus("connected", 0, timeout = 5): 
        bailOut(msg = "Disconnect failed\nExiting", disconnect = False)
    
    c = t2tCmd()
    c.cmd = "CLEANUP"
    cmdEx(c)
    
    print "End.\n"
    
    t2tOutputFileName(None, None)

except Exception, e:
    msg =  "--------------- UNEXPECTED ERROR ---------------\n" + e.__str__()
    msg += "\n------------------------------------------------\n"
    msg += "\nThe application has been forcely closed!\n"
    bailOut(msg = msg, disconnect = tobii_connected)
    

File: TobiiExampleScript/Tobii_helpers/t2t/examples/perl/followEyes.pl
#!/usr/bin/env perl -w
##
# Luca Filippin - July 2010 - luca.filippin@gmail.com                                                
# Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste. All rights reserved.
# 
# PLEASE HAVE A LOOK TO THE SWIG DOCUMENTATION FOR A DEEPER UNDERSTANDING OF THIS CODE
# NOTE: Swig wrapper to t2t makes use of shadow classes and swig helpers (see t2tsw.i)
##

#use strict;
use Getopt::Long;
use File::Basename;
use t2tHelpers;
use Tk::CursorControl;
use Tk;

GetOptions(	"help|h"=>\$opt_help,
			"version"=>\$opt_version,
			"log-file|l:s"=>\$opt_log_file,
   			"port|p:i"=>\$opt_tobii_port,
            "calibration-file|c:s"=>\$opt_calibration_file,
            "tracking-on|t"=>\$opt_tracking_on,
            "camera|m"=>\$opt_camera,
            "window-width|w:f"=>\$opt_window_width,
            "window-height|e:f"=>\$opt_window_height,
            "quick-start|q"=>\$opt_quick_start,
            "debug-mode|d"=>\$opt_debug_mode,
            "validity|v:i"=>\$opt_validity);
            
($command) = fileparse($0);

$version=<<VERSION;
+-------------------------------------------------------------------------------------+
|  followEyes.pl v1.0                                                                 |
|  This is a sample application which shows eyes movements as tracked by the Tobii ET |
|-------------------------------------------------------------------------------------|
|  Luca Filippin - July 2010 - luca.filippin\@gmail.com                                |	
|  Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste     |
+-------------------------------------------------------------------------------------+
VERSION

$usage=<<USAGE;

Usage: $command [options] <TET ip address>

Display eyes movements by traking them through a Tobii Eyes Tracker

Options:
  -h, --help            show this help message and exit
  --version
  -l TOBII_LOG, --log-file TOBII_LOG
                        log file name
  -c CALIBRATION_FILE, --calibration-file CALIBRATION_FILE
                        Use a stored calibration file
  -p TOBII_PORT, --port TOBII_PORT
                        TET server listening port
  -d, --debug-mode      Print gaze samples
  -m, --camera          Use camera coordinates
  -q, --quick-start     Start soon after the command has been launched
  -w WINDOW_WIDTH, --window-width WINDOW_WIDTH
                        Window width: range (0,1]
  -e WINDOW_HEIGHT, --window-height WINDOW_HEIGHT
                        Window heigth: range (0,1]
  -t, --tracking-on     Show the path made by the turtle
  -v {0,1,2,3,4}, --validity {0,1,2,3,4}
                        validity level: 0 = certainly, 1 = probably, 2 = 50%,
                        3 = likely not 4 = surely not
USAGE

if ($opt_version) {
	print "$version";
	exit 0;
}

if (@ARGV < 1 or $opt_help) {
  	print "$usage";
  	exit 0;
}

if (!defined($opt_tobii_port)) { $opt_tobii_port=4455; }
if (!defined($opt_validity)) { $opt_validity=2; }
if (!defined($opt_window_width)) { $opt_window_width=0.5; }
if (!defined($opt_window_height)) { $opt_window_height=0.5; }

if ($opt_window_height <= 0 || $opt_window_height > 1) { 
	die "Bad window length value: $opt_window_height. Must be in (0,1]"; 
}
if ($opt_window_width <= 0 || $opt_window_width > 1) { 
	die "Bad window length value: $opt_window_width.  Must be in (0,1]"; 
}

$opt_tobii_ip = $ARGV[0];

sub banner() {
	print "$version\n";
}

sub follow_eyes {
	my ($win, $canvas, $p_stop, $validity, $camera, $debug_on, $track_on,  $w, $h, $sx, $sy, $to_delete) = @_;
	my ($move, $msg, $x, $y) = (1, "", 0, 0);
	
	my $c = new t2tsw::t2tCmd();
    $c->{cmd} = "GET_SAMPLE_EXT";
    my $smp = t2tHelpers::cmdEx($c)->{sample_ext};

	if (defined $to_delete) {
		foreach my $o (@{$to_delete}) {
			$canvas->delete($o);
		}
		$canvas->update;
	} 
	
	my ($lx, $ly, $rx, $ry);
	
	if ($camera) {
		$lx = 1 - $smp->{lxcam};
		$ly = $smp->{lycam};
		$rx = 1 - $smp->{rxcam};
		$ry = $smp->{rycam};
	} else {
		$lx = $smp->{lx};
		$ly = $smp->{ly};
		$rx = $smp->{rx};
		$ry = $smp->{ry};
	}
	
	$to_delete = []; 
	
	if ($lx >= 0 && $ly >= 0 && $rx >= 0 && $ry >= 0 && $smp->{lval} <= $validity && $smp->{rval} <= $validity) {
		$m = sprintf("Sample OO: %s", $smp->str());
		$x = ($lx + $rx)*$w/2;
		$y = ($ly + $ry)*$h/2;
	}
	elsif ($rx >= 0 && $ry >= 0 && $smp->{rval} <= $validity) {
		$m = sprintf("Sample OX: %s", $smp->str());
		$x = $rx*$w;
		$y = $ry*$h;
		# just copy over because we draw always both
		$lx = $rx;
		$ly = $ry;
		$smp->{lval} = $smp->{rval};
	}
	elsif ($lx >= 0 && $ly >= 0 && $smp->{lval} <= $validity) {
		$m = sprintf("Sample XO: %s", $smp->str());
		$x = $lx*$w;
		$y = $ly*$h;
		# just copy over because we draw always both
		$rx = $lx;
		$ry = $ly;
		$smp->{rval} = $smp->{lval}; 
	}
	else {
		$m = sprintf("Sample XX: %s", $smp->str());
		$x = $sx;
		$y = $sy;
		$move = 0;
	}
	if ($move) {
		my @colors = ("red", "orange", "yellow", "blue", "white");
		my $osx = $canvas->createOval($lx*$w - 5, $ly*$h - 5, $lx*$w + 5, $ly*$h + 5, -fill => $colors[$smp->{lval}]);
        my $odx = $canvas->createOval($rx*$w - 5, $ry*$h - 5, $rx*$w + 5, $ry*$h + 5, -fill => $colors[$smp->{rval}]);
		if  ($track_on) {
 			$canvas->createLine($sx, $sy, $x, $y, -width => 1, -fill => "green");
        } else {
        	push @$to_delete, $osx;
        	push @$to_delete, $odx;
        }
        $canvas->update;
        $m .= sprintf("\nMove to (x, y) = (%.0f, %.0f)", $x, $y); 
	} else {
		$m .= "\nPrevious sample SKIPPED";
	}
	
	printf("$m\n") if ($debug_on);
	
	if (! $$p_stop) {
		$canvas->after(10, \&follow_eyes, $win, $canvas, $p_stop, $validity, $camera, $debug_on, $track_on,  $w, $h, $x, $y, $to_delete);
	} else {
		$win->destroy();
	}
}

sub _get_out {
	$stop = 1;
}

eval {
	$disconnect = 0;
	banner();
	t2tsw::t2tOutputFileName($opt_log_file, "w") if defined $opt_log_file;
	
	# Connect to the tobii
    print "Connecting to the TET server @".$opt_tobii_ip.":".$opt_tobii_port."...\n";
    my $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CONNECT";
    $c->{prm}->{connect}->{ip_address} = $opt_tobii_ip;
    $c->{prm}->{connect}->{port} = $opt_tobii_port;
    t2tHelpers::cmdEx($c);
	t2tHelpers::checkStatus('connect', 1, 5);
	t2tHelpers::bailOut("Connection failed\nExiting\n", 0) unless t2tHelpers::checkStatus('connected', 1, 5);
	$disconnect = 1;
    
    if (defined $opt_calibration_file) {
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "START_CALIBRATION";
    	$c->{prm}->{start_calibration}->{load_from_file} = 1;
    	$c->{prm}->{start_calibration}->{cmatrix}->{cols} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{rows} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{vals} = undef;
    	$c->{prm}->{start_calibration}->{fname} = $opt_calibration_file;
    	t2tHelpers::cmdEx($c);
    	
    	t2tHelpers::checkStatus('calibrating', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 0, 5);
    } 
    
    if ($opt_quick_start) {
    	print "**** QUIT THE APPEARING WINDOW TO STOP *****\n";
    } else {
    	t2tHelpers::rawInput("\n***** PRESS ANY KEY TO START TRACKING AND QUIT THE APPEARING WINDOW TO STOP *****\n");
    }
    
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "START_TRACKING";
    t2tHelpers::cmdEx($c);
    t2tHelpers::checkStatus('running', 1, 5);
    t2tHelpers::bailOut("Start tracking failed\nExiting\n", 1) unless t2tHelpers::checkStatus('runstarted', 1, 5);
    
    # Tk initialisation
	my $window = MainWindow->new;
	$window->configure(-title => "Follow the eyes movements through a Tobii eyetracker");
	$window->CursorControl->hide($window);
	my $w_height = $window->screenheight() * $opt_window_height;
	my $w_width = $window->screenwidth() * $opt_window_width;
	$window->configure( -width=>$w_width, -height=>$w_height );
	$window->resizable(0, 0);
	my $canvas = $window->Canvas(-background=>"black", -width=>$w_width, -height=>$w_height, -borderwidth => 0);	
	$canvas->pack(-expand => 1, -fill => 'both');
	$stop = 0;
    follow_eyes($window, $canvas, \$stop, $opt_validity, $opt_camera, $opt_debug_mode, $opt_tracking_on,  $w_width, $w_height, $w_width/2, $w_height/2);
    $window->protocol('WM_DELETE_WINDOW',\&_get_out);
	MainLoop;
	
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "STOP_TRACKING";
    t2tHelpers::cmdEx($c);
    t2tHelpers::checkStatus('stop', 1, 5);
    t2tHelpers::bailOut("Stop tracking failed\nExiting\n", 0) unless t2tHelpers::checkStatus('running', 0, 5);
    
    $disconnect = 0;;
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "DISCONNECT";
    t2tHelpers::cmdEx($c, 0);
    t2tHelpers::checkStatus('disconnect', 1, 5);
	t2tHelpers::bailOut("Disconnection failed\nExiting\n", 0) unless t2tHelpers::checkStatus('connected', 0, 5);
	
	$c = new t2tsw::t2tCmd();
    $c->{cmd} = "CLEANUP";
    t2tHelpers::cmdEx($c);
    
	printf "End.\n";
	
} or do {
	my $msg = "--------------- UNEXPECTED ERROR ---------------\n$@";
    $msg .= "------------------------------------------------\n";
    $msg .= "The application has been forcely closed!\n";
    t2tHelpers::bailOut($msg, $disconnect);
}
